inferno-lsp/src/Inferno/LSP/Server.hs

```diff
@@ -212,6 +204,29 @@ sendDiagnostics :: J.NormalizedUri -> J.TextDocumentVersion -> [J.Diagnostic] ->
 sendDiagnostics fileUri version diags =
   publishDiagnostics 100 fileUri version (partitionBySource diags)
 
+parseAndInferWithTimeout ::
+  forall m1 m2 c.
+  (MonadIO m2, MonadThrow m1, Pretty c, Eq c) =>
+  ((UUID, UTCTime) -> IO ()) ->
+  ((UUID, UTCTime) -> ParsedResult -> IO ParsedResult) ->
+  ModuleMap m1 c ->
+  [Maybe Ident] ->
+  T.Text ->
+  (InfernoType -> Either T.Text ()) ->
+  m2 ParsedResult
+parseAndInferWithTimeout beforeParse afterParse prelude idents doc_txt validateInput = do
+  ts <- liftIO getCurrentTime
+  uuid <- UUID <$> liftIO UUID.V4.nextRandom
+
+  liftIO $ beforeParse (uuid, ts)
+  result <- do
+    -- Timeout parsing and type checking after 10 seconds
+    mResult <- liftIO $ timeout (10 * 1000000) $ parseAndInfer @m1 @_ @c prelude idents doc_txt validateInput
```

Comment:
I recently learnt about `NumericUnderscores` which is great for readability of these big constants, lets you write ten million as `10_000_000`.

Reply:
Ah, that's very nice, thanks :)